### ISSUE TITLE
[3.x] ConnectionInterface & support for NoSQL Proposal

### ISSUE THREAD

**Issue by timhoeppner** [cakephp/cakephp#8148]
## Intro

Before I begin, I just want to say I don't claim to be an expert in CakePHP, Relational or Document model databases and it's very possible I've missed big issues in what I'm proposing. With that being said, I'm a huge fan of CakePHP and what it lets me do personally; my only interest here is contributing and playing a part in the community.

I've been working integrating MongoDB into CakePHP and it's come to my attention that the current database model in 3.x is heavily dependant on SQL. After doing my own research and coming up with an idea of how I might change things, I also came across https://github.com/cakephp/cakephp/issues/2617 which talks about some of the same issues I'm coming across. Although I've completed what I need for the time being it is somewhat incomplete due to how CakePHP's model of Connections and Drivers currently are.

I also just want to mention that my specific implementation doesn't replace relational databases instead uses them to store only certain types of data, the rest remains in relational tables. I've read some comments about no interest in NoSQL databases for CakePHP but I'm thinking that's because the feeling is that people want to use them to replace relational databases. Considering that NoSQL databases lack schema this doesn't really fit in the beautiful ORM that has been built for CakePHP but rather it's a completely separate way of storing and accessing data.
## Issues

ConnectionInterface is interesting, the concrete implementation class is Connection. What I've noticed throughout the code base is that ConnectionInterface is what's passed around but the functions using the interface are using other functions from Connection which aren't a part of ConnectionInterface. I've compiled a list of functions which are called from ConnectionInterface objects (**bold** functions actually belong to the interface):
- **configName()**
- **config()**
- **transactional()**
- **disableConstraints()**
- **logQueries()**
- driver()
- execute()
- run()
- compileQuery
- schemaCollection()
- prepare()
- newQuery()
- inTransaction()

If there are eventual other implementations of ConnectionInterface all these other methods should be a part of the interface otherwise other ConnectionInterface objects won't be compatible. This is awkward. Calling functions outside the passed in interface also breaks type-hinting. Just to be clear I'm talking about when a function prototype is something like `public function createSql(ConnectionInterface $connection)` and then calling `$connection->driver()` which isn't a part of ConnectionInterface.
### ConnectionInterface and Connection

ConnectionInterface should remain nice and lean but be changed as follows:
- Remove both transactional() and disableConstraints() and move these into the driver as these both feel like driver level functions
- Add driver()

All of the query related functions in Connection could be moved to driver as that seems more appropriate as they don't seem directly applicable to a connection.
### Driver abstraction

The abstract class Driver could provide an additional layer of abstraction to make room for other types of drivers that are not SQL based. Any SQL related functions could be moved onto say

`abstract class SqlDriver extends Driver`

and then all SQL related drivers would extend that. If functionality requires SQL (basically all of it currently) then SqlDriver would be passed instead of the more generic Driver.
## Summary

That's about as far as I've dug down. I was originally going to work on the implementation and submit a pull request but I think that I would have a hard time getting it accepted before having a discussion.

**Comment by josegonzalez** [comment]
Note: with the current implementation, we already have support for [Elasticsearch](https://github.com/cakephp/elastic-search).

**Comment by markstory** [comment]
Removing methods is not an option as it will break compatibility. Some nonrelational stores do support transactions, so I am not clear on why transactional() would need to be removed.

I don't think its necessary to even use the driver classes in a nosql implementation. We were able to side step that issue entirely with the elasticsearch plugin.

**Comment by timhoeppner** [comment]
@josegonzalez 
Thanks for the note. As I mentioned I did have success with integrating what I needed in my application with MongoDB, I'm just thinking these changes **could** help with other third-party developers wanting to do something similar.

@markstory 
It's not clear to me the distinction between connections and drivers in CakePHP. My suggestion of **moving** transactional() to the driver would be so that anything query related would be in the driver instead of the connection but maybe that isn't the design intent.

What's the idea behind ConnectionInterface? Everything that receives it just assumes it's a Connection object? Is there interest in addressing what I've identified here? I understand the need for backward compatibility and there are certainly strategies to address the concern if there's interest to move forward.

**Comment by josegonzalez** [comment]
@timhoeppner it might help us conceptualize your changes if you could post your mongodb plugin (even if it is a WIP) and we can see where this might help.

**Comment by ionas** [comment]
I like @lorenzo statement that an ORM should not mix sql and non-sqlish interfaces. There is ElasticSearch to look at and if I had need for a Key/Value, Document or Graph storage I'd look at how Elastic Search is done.


Going the 2.x way and:
- losing on options, flexibility and power for non-sqlish datasources
- losing on options and power for sqlish datasources
just to have one single interface (like Litihum still tries) was a bad approach in cake2 and is a bad approach still.

In general as far as I understood, the logic of Entities including validation can be reused, you will however need to re-implement business logic around the "structure" (currently called rules for the sqlish ORM) and database data. E.g. field constraints (validations) and transformations (entity setters and getters) should work well. I am not sure how much work you need to do to implement something like newEntity, patchEntity etc. but that's imho the way to go.

Good luck.

**Comment by timhoeppner** [comment]
@josegonzalez Its not something I can share as its something I developed during my 8-5 job otherwise I'd be happy too. I'll just stick with what I have and keep hacking away. Thanks for your consideration.

**Comment by josegonzalez** [comment]
Hmm okay. What underlying library are you using for the connection, and do you have a list of features you want to support? Maybe I can write a simple implementation and we can see what pain points we have?

**Comment by timhoeppner** [comment]
Sure that would be great, I'm using [this library](https://github.com/mongodb/mongo-php-library).

My currently implementation has the following features:
 - Connection can be managed using ConnectionManager with a separate database configuration in config/app.php
 - Runs in parallel with current ORM (I'm not trying to replace this) built on a Collection class which would be somewhat synonymous to the Table class
 - By default a model extended from the Collection class, for example TestsCollection, would map to the test collection inside the database defined in config/app.php. This can be overridden using `$collection->collection("alternate_name")`.
 - At this point I've just implemented __call in my Collection class which maps functions onto \MongoDB\Collection

Future features:
 - Integrate all the query builder functionality
 - I don't want/need table associations, again I'm not trying to use MongoDB as a relational database
 - Right now the collection object returns back whatever the underlying mapped MongoDB library function returns. I would like it to be more consistent with how CakePHP returns results.
 - Include a similar interface for loading the collection classes like TableRegistry::get() into whatever requires them.


I just want to emphasize, I'm not a CakePHP expert and my proposal was trying to address what I had seen with connections/drivers. Since I haven't had a chance to look at the query builder (and other related functionality) I have no idea on the effort there will be there. I'll be taking a closer looks at how you guys did elasticsearch over the next few days and see if there's some quick lessons I can learn there.

**Comment by markstory** [comment]
> What's the idea behind ConnectionInterface? Everything that receives it just assumes it's a Connection object? Is there interest in addressing what I've identified here? I understand the need for backward compatibility and there are certainly strategies to address the concern if there's interest to move forward.

The Connection/Driver separation is a bit sql-esque as well. The connection object is intended to represent the 'connection' at a high-level. It manages tasks like:

* Configuration
* Driver selection (more on that in a bit)
* connection state
* Basic query operations (begin, commit, insert, update etc)

By necessity this brings a lot of SQL baggage, and we've not included those methods in the interface as many of the methods defined in `Database\Connection` are meaningless in a nosql context. Drivers contain platform/vendor specific specializations related to the dialect of SQL that each vendor provides. They also provide the option to have abstract around native procedural interfaces and PDO.

**Comment by rfenner** [comment]
I wrote a plugin for mongo back when 3.x was still in beta when i first started using it. It was based essentially a port of the existing system minus the sql pieces that didn't seem to need to be in there.
I felt at the time that the driver and connections and types could probably be generalized a little more to remove the SQL specific stuff but I didn't want to modify the source at the time so I just created the plugin.
[Cakephp3.x-mongodb](https://github.com/rfenner/cakephp3.x-mongodb/tree/master/NoSql)
I just recently updated it to support the new php driver.
It works but it's not super tested. But it should give an idea of what you were saying about the generalization of the connecton and drivers.
